### PR TITLE
[Parse/Sema] Diagnose invalid attributes for ParamDecl and GenericTypeParamDecl

### PR DESCRIPTION
**Pull request by rintaro**
Previously, invalid attributes were silently accepted and ignored.

https://bugs.swift.org/browse/SR-8202

### CODE REVIEW

**Comment by rintaro** [comment]
@DougGregor @jrose-apple Could you take a look?

@DougGregor `@available` at generic parameters is really needed and actually works?
```swift
class B<@available(*, unavailable, renamed: "U") T> {}
extension B where T: Equatable { }
```
This can be compiled, but I don't think it should be. 
[Language Reference](https://docs.swift.org/swift-book/ReferenceManual/GenericParametersAndArguments.html#//appleref/swift/grammar/generic-parameter-clause) does not allow them.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Review by jrose-apple — APPROVED** [review]
Doug's out on vacation. I agree that generic parameters should not support `@available`, though. They really have to be as available as the base type, at least today.

We will want to do a source compat run before merging this.

**Comment by rintaro** [comment]
@swift-ci Please Test Source Compatibility

**Comment by jrose-apple** [comment]
Linux issues have been fixed.

@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
@swift-ci Please test source compatibility

**Comment by rintaro** [comment]
@swift-ci Please some test Linux

**Comment by rintaro** [comment]
@swift-ci Please some test Linux

**Comment by rintaro** [comment]
@swift-ci Please some test Linux

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
@swift-ci Please test source compatibility

**Comment by rintaro** [comment]
holding off until source-combat-suite become healthy.

**Comment by rintaro** [comment]
@swift-ci Please test source compatibility

**Pull request merged**
